include/re_atomic.h

```diff
@@ -7,40 +7,346 @@
 #ifndef RE_H_ATOMIC__
 #define RE_H_ATOMIC__
 
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 /* C11 */
 #if defined(HAVE_ATOMIC) && !defined(__STDC_NO_ATOMICS__)
+
 #include <stdatomic.h>
 
-/* With MSVC volatile is atomic */
-#elif defined(_MSC_VER)
-#define RE_ATOMIC volatile
```

Comment:
RE_ATOMIC definition for MSVC is missing now.

Reply:
`volatile` is not atomic, even on MSVC. It does not support `_Atomic`, so `RE_ATOMIC` is rightfully defined to empty. Atomicity and memory ordering are implemented through intrinsics.

Reply:
We have to keep the volatile keyword, since we rely on this for some simple shared thread run booleans.

https://docs.microsoft.com/en-us/cpp/cpp/volatile-cpp?view=msvc-170#end-of-iso-conformant

Reply:
The mode is "/volatile:ms" since VS 2005 for x86 "default when architectures other than ARM are targeted" we can add /volatile:ms for explicit  MSVC arm support (but nothing we test or support at the moment). The volatile keyword is only used for MSVC RE_ATOMIC fallback and every volatile write has release semantics and every volatile read has acquire semantics.

Of course for other compilers, `volatile` should be avoided as atomic replacement but visual studio is a valid exception.

Ref: https://stackoverflow.com/a/44724954

Reply:
As I said before, memory ordering implemented for `volatile` by MSVC does not make it an atomic. Marking variables `volatile` for inter-thread communication is incorrect, even on MSVC. You can trick the compiler to generate the code that will work correctly in multithreaded environment on a given target architecture (like using simple loads or stores on small variables on x86; not even that on larger ones, like `uint64_t`, which may get split on 32-bit x86), but in general `volatile` does not give you atomicity. For example, `++i` is not atomic if `i` is `volatile`, regardless of its size or target architecture or compiler.

BTW, if I remember correctly (I haven't looked thoroughly yet), the uses of `RE_ATOMIC` in `main` actually don't care about memory ordering aspect of the atomics. That is, all loads and stores there could use `memory_order_relaxed`. The only reason to use atomics there is to ensure atomicity, which is to force the compiler use the correct instructions for loads and stores. On x86 that makes little difference with regular loads and stores (it mostly restricts the compiler from doing weird stuff like `push` to write to a variable), but it may make a more meaningful difference on other archs.

Reply:
Yes the code comment is misleading, we should change this:

```diff
+/* volatile is not atomic but ensures read/write are not optimized away (on MSVC this implies memory ordering) */
+#define RE_ATOMIC volatile
```

`volatile` itself is of course not atomic but ensures the boolean is not optimized away (`_Atomic` implies that too I think) and guarantees memory ordering on MSVC. The Interlocked API also uses `volatile`:

```c
LONG InterlockedIncrement(
  [in, out] LONG volatile *Addend
);
```

Reply:
WinAPI is actually not used and not present on x64. But Microsoft was not consistent about `volatile` either, as e.g. on WinCE these functions were declared without `volatile`.

In any cases, `volatile` is not needed if you access atomics using `re_atomic_*` API - which you should anyway on any compiler that does not support C11 `_Atomic`. So, I left `RE_ATOMIC` empty for MSVC and updated the atomic uses in `main`.